pkg/kv/kvserver/replica_rankings_test.go

```diff
@@ -73,3 +80,91 @@ func TestReplicaRankings(t *testing.T) {
+func TestAddSSTQPSStat(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	tc := serverutils.StartNewTestCluster(t, 1, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+	})
+
+	defer tc.Stopper().Stop(ctx)
+	ts := tc.Server(0)
+	db := ts.DB()
+	conn := tc.ServerConn(0)
+	sqlDB := sqlutils.MakeSQLRunner(conn)
+
+	// Construct an sst with 100 keys that will be reused with different divisors.
+	sstKeys := make([]sstutil.KV, 100)
+	for i := range sstKeys {
+		sstKeys[i] = sstutil.KV{fmt.Sprintf("%3d", i+1), 1, "value"}
+	}
+	sst, start, end := sstutil.MakeSST(t, sstKeys)
+	requestSize := float64(len(sst))
+
+	sstReq := &roachpb.AddSSTableRequest{
+		RequestHeader: roachpb.RequestHeader{Key: start, EndKey: end},
+		Data:          sst,
+		MVCCStats:     sstutil.ComputeStats(t, sst),
+	}
+
+	get := &roachpb.GetRequest{
+		RequestHeader: roachpb.RequestHeader{Key: roachpb.Key("001")},
+	}
+
+	addSSTBA := roachpb.BatchRequest{}
+	nonSSTBA := roachpb.BatchRequest{}
+	addSSTBA.Add(sstReq)
+	nonSSTBA.Add(get)
+
+	// When the factor is set to 0, it is disabled and we expect uniform 1 QPS.
+	// In all other cases, we expect 1 + the size of a
+	// AddSSTableRequest/factor. If no AddSStableRequest exists within the
+	// request, it should be cost 1, regardless of factor.
+	testCases := []struct {
+		addsstRequestFactor int
+		expectedQPS         float64
+		ba                  roachpb.BatchRequest
+	}{
+		{0, 1, addSSTBA},
+		{100, 1, nonSSTBA},
+		{1, 1 + requestSize, addSSTBA},
+		{10, 1 + requestSize/10, addSSTBA},
+		{50, 1 + requestSize/50, addSSTBA},
+		{100, 1 + requestSize/100, addSSTBA},
+	}
+
+	// Send an AddSSTRequest once to create the key range.
+	_, pErr := db.NonTransactionalSender().Send(ctx, addSSTBA)
+	require.Nil(t, pErr)
+
+	store, err := ts.GetStores().(*Stores).GetStore(ts.GetFirstStoreID())
+	require.NoError(t, err)
+
+	repl := store.LookupReplica(roachpb.RKey(start))
+	require.NotNil(t, repl)
+
+	for _, testCase := range testCases {
+		sqlDB.Exec(t, fmt.Sprintf(`SET CLUSTER setting kv.replica_stats.addsst_request_size_factor = %d`, testCase.addsstRequestFactor))
+
+		qpsBefore, durationBefore := repl.QueriesPerSecond()
+		queriesBefore := qpsBefore * durationBefore.Seconds()
+
+		_, pErr = db.NonTransactionalSender().Send(ctx, testCase.ba)
+		require.Nil(t, pErr)
+
+		qpsAfter, durationAfter := repl.QueriesPerSecond()
+		queriesAfter := qpsAfter * durationAfter.Seconds()
+		queriesIncrease := queriesAfter - queriesBefore
+
+		// If queries are correctly recorded, we should see increase in query count by the expected
+		// QPS. However, it is possible to to get a slightly higher or lower number, due to rounding
+		// or timing errors; the result is calculated using floating point division. We allow a
+		// tolerance of 4 to avoid flaking the test due to this inaccuracy.
+		require.InDelta(t, queriesIncrease, testCase.expectedQPS, 4)
```

Comment:
Expose a method testGetQPSSum to assert exactly on the QPS recorded.

Reply:
Exposed method, however here it doesn't seem possible to avoid the issue with inaccuracy. Investigating the cause further.